Feature: Add Dash/Sprint Shoe Dash movement options

kielbasiago opened the pull request:
Reworked auto-sprint and the dash flag introduced in [#22 Feature/sprint-shoes-b-dash](https://github.com/AtmaTek/WorldsCollide/pull/22) as when I began adding a "b dash" option (w/o sprint shoes) to a it was cumbersome.

## Changes

### Added Sprint Shoes as an objective result
- Makes original move speed (no auto sprint) a more viable option by adding deterministic access to sprint shoes
- Can be used with newly added "sprint shoes dash" option
 
### Added `-noshoes` flag
- Removes shoes from appearing in shops, chests, etc.
- Just allows a more strategic use of sprint shoes with the objective result
- Sprint shoes are a dead item in half of the movement options

### Reworked `--auto-sprint` into `--movement`
- Removed `--auto-sprint` flag
- Added `--movement` flag
  - Options listed below

### `--movement` options
**Original**
- Move speed 2 by default
- Move speed 3 with sprint shoes on active party
- This makes no changes to the rom

**Auto Sprint**
- This is equivalent to `--auto-sprint` flag
- Move speed 3 by default
- Move speed 2 when holding B

**B-Dash**
- Move speed 3 by default
- Move speed 4 when holding B

**Sprint Shoes B-Dash**
- Move speed 3 by default
- Move speed 2 when holding B
- Move speed 4 when:
  - Sprint shoes equipped on active party 
  - Holding B
- This is equivalent to the `--shoe-dash` flag from the old PR

## Notes
- Sliding dash fix was ported from SilentEnigma's Sliding Dash v1.1 fix: https://silentenigma.neocities.org/ff6/index.html

kielbasiago committed:
Feature: Add movement options, b-dash, sprint shoe options

kielbasiago commented on settings/movement.py:
```diff
@@ -0,0 +1,182 @@
+        # moving at dash speed in Owzer's door room, or carrying it out via the door glitch will cause graphical artifacting randomly.
+        # Simply disabling B button in Owzers to keep it consistent in WC. Will not worry about the door glitch
+        src = [
+            # Need a sanity check here - This is causing an issue. Go to most maps and you will be unable to move south
```
Had trouble getting this to work. Not sure if it ever even worked to be honest.. But I'm a bit out of ideas on what to do here. 

This is only a light graphical bug as the scrambling/artifacting is localized to the room it occurs in. i.e. once you escape the door room everything is back to normal. But it happens enough to where i'd probably not put in dash without this

kielbasiago commented on data/movement.py:
```diff
@@ -0,0 +1,11 @@
+class MovementActions:
```
need to change these here for readability / understandability's sake
```
ORIGINAL = `o`
AUTO_SPRINT = 'as'
B_DASH = 'bd'
SPRINT_SHOES_B_DASH = 'ssbd'
```

asilverthorn commented on args/misc.py:
```diff
@@ -1,10 +1,15 @@
 def parse(parser):
     misc = parser.add_argument_group("Misc.")
-    misc.add_argument("-as", "--auto-sprint", action = "store_true",
-                      help = "Player always sprints. Sprint Shoes have no effect")
```
Would it be possible to leave it in for now, but just have it call the other flag? Removing this option would break every saved seedbot flag which uses the -as flag. At some point (maybe when the major release happens?) we'd want to remove these deprecated flags, but for now cleaning up all of the existing flagsets would prove problematic.

wrjones104 replied:
I could just do a find & replace on the preset file if we went this way - I don't think it would be a huge lift

kielbasiago replied:
Added the argument back, and going to completely ignore it..

in settings/movement.py we have `self.movement = args.movement or MovementActions.AUTO_SPRINT` which means we will use auto sprint whether or not `-as` was ever specified

This means that vanilla movement behavior must be opt-in. I think that's probably better for the randomizer overall as it is a huge QOL (and the applications of forcing player to walk are probably very niche)

kielbasiago replied:
(re-added on line 23)

asilverthorn commented on data/movement.py:
```diff
@@ -0,0 +1,11 @@
+    ORIGINAL = 'og'
+    # Auto sprint, holding B with sprint will move player at ms 3
+    AUTO_SPRINT = 'as'
+    # Holding B with sprint will move player at ms 5
```
Minor: The B_DASH and SPRINT_SHOES_B_DASH comments are the same.

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,169 @@
+        # moving at dash speed in Owzer's door room, or carrying it out via the door glitch will cause graphical artifacting randomly.
+        # Simply disabling B button in Owzers to keep it consistent in WC. Will not worry about the door glitch
+        src = [
+            # Need a sanity check here - This is causing an issue. Go to most maps and you will be unable to move south
```
I'll take a look at this; I think we'll want this fixed before merging to main.

asilverthorn replied:
Which maps? I enabled this (with changing the Allocate call to the Write call, since I didnt' want to recalculate the length) and didn't experience this issue.

kielbasiago replied:
Sorry, specifically I have noticed it in Albrook and Imperial base (before sealed gate)

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,169 @@
+        length = 0
+        src = []
+
+        (length, src) = self.get_auto_sprint_src()
```
Rather than calculate the length yourself, you should be able to just call 

space = Write(Bank.F0, src, "Sprint subroutine")

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,169 @@
+        src = [
+            asm.JSL(space.start_address_snes),
+        ]
+        space = Reserve(0x04e21, 0x04e37, "auto sprint", asm.NOP())
```
Looking at this vanilla code, I see that it starts with this:

```
C0/4E21:	ADB81E  	LDA $1EB8      (load event bits)
C0/4E24:	2902    	AND #$02       (is "disable sprinting" set?)
C0/4E26:	D00B    	BNE $4E33      (branch if so)
```

I'm not sure in what cases the "disable sprinting" flag would otherwise be set, but I think we'd want to keep that logic in place.

kielbasiago replied:
this space is reserving the same as main right now - https://github.com/AtmaTek/WorldsCollide/blob/main/settings/auto_sprint.py#L35

Not sure how we want to handle that

asilverthorn replied:
oh, thanks. No need to change then.

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,169 @@
+        FIELD_RAM_SPEED = 0x0875
+
+        # moving at dash speed in Owzer's door room, or carrying it out via the door glitch will cause graphical artifacting randomly.
+        # Simply disabling B button in Owzers to keep it consistent in WC. Will not worry about the door glitch
```
What's the door glitch?

kielbasiago replied:
You can carry the owzer's door effect outside of the mansion by opening the menu at certain times:
[Drew doing it for his glitched runs](https://www.twitch.tv/videos/1457261894?t=0h5m37s)

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,169 @@
+    def get_auto_sprint_src(self):
+        import args
+        CURRENT_MAP_BYTE = 0x82 # 2 bytes
+        OWZERS_MANSION_ID = 0x0D1 # the door room can create visual artifacts on the map while dashing
```
I think this should be the Owzer's Basement map which, if I'm looking at https://everything8215.github.io/ff6tools/ff6tools.html correctly, should be 207 (0x0cf)

asilverthorn replied:
I enabled this and confirmed it should be 0xCF.

![image](https://user-images.githubusercontent.com/96998881/166614229-630dd250-850e-47bf-affc-2ccc8d538c73.png)

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,169 @@
+from data.movement import MovementActions
+from memory.space import Allocate, Bank, Reserve, Read
```
Read is unused, can remove

asilverthorn commented on settings/movement.py:
```diff
@@ -0,0 +1,163 @@
+            # asm.LDA(CURRENT_MAP_BYTE, asm.ABS),         # if current map owzers mansion, disable the b-button
+            # asm.CMP(OWZERS_MANSION_ID, asm.IMM16),
+            # asm.BEQ("STORE_DEFAULT"),
+            # asm.A8(),
```
Add this line before this asm.A8() to fix the issue:

`asm.LDA(0x0000, asm.IMM16),                 # clear A`

The issue is that A, on some maps, has it's upper byte set. That's mucking up some downstream use of it.

asilverthorn commented on data/movement.py:
```diff
@@ -0,0 +1,25 @@
+    SPRINT = 3  # Original FF6 sprint speed
+    DASH = 4    # Custom, move twice as fast as sprint.
+
+class MovementActions(IntEnum):
```
I get this error when generating a seed:

```
  File "C:\Users\mtkm\WorldsCollide\data\movement.py", line 10, in <module>
    class MovementActions(IntEnum):
  File "C:\Users\mtkm\AppData\Local\Programs\Python\Python310\lib\enum.py", line 289, in __new__
    enum_member = __new__(enum_class, *args)
ValueError: invalid literal for int() with base 10: 'og'
```

asilverthorn replied:
I suspect you meant to put (IntEnum) in the MovementSpeed class instead

kielbasiago commented:
when no movement options specificied it shows "none" in the menu https://i.imgur.com/o7EvF7c.png

kielbasiago commented:
Pushed an update: 
- Now properly display the title (was showing key `og`, `as`, `bd`, `ssbd` but now shows a readable name.
- No longer shows "None" when no movement is specified.
- `-as` is no longer mutually exclusive with `-move`. Instead we are now completely ignoring the flag.
  - I can only remember one time I've seen others play original speed and it was the living seed 
 
## Menu
**longest label (sprint shoes b dash) to confirm line length is not too long**
Building with `python3 ./wc.py -i ff3.smc -as -move ssbd`

![image](https://user-images.githubusercontent.com/95580337/170723766-ebe7df84-e667-44ba-bbf7-bb8fb9ce0750.png)

---------
**defaulting to auto sprint**
Building with `python3 ./wc.py -i ff3.smc`
![image](https://user-images.githubusercontent.com/95580337/170724834-5dc59fdc-4e29-46e8-aa14-71cd81be07c4.png)

asilverthorn committed:
Adding Auto Life 3 and Auto Dog Block objective results (#36)

asilverthorn committed:
Stray Flash mod (#33)

asilverthorn committed:
Adding Osteoclave's bugfix (#29)

asilverthorn committed:
Adding expensive rods and super balls flags (#27)

asilverthorn committed:
Bugfix: Fix weapon special abilities with Capture and Multi-steals only giving 1 item (#25)

* Porting Assassin's Capture bugfix

* Adding Bropedio's Multi-Steal fix

* Adding multi-steal dialog fix

* updating flag doc

* Small reduction in C1 usage

* Fixing bug with -fc that causes escape from battle (and possibly other actions) to crash

asilverthorn committed:
Feature: Flashes: Moving flashes to Graphics; Slowing scrolling background during Air Force fight and Baren falls; add minimap high contrast option (#23)

* Removing scrolling background during Air Force fight

* switching to slowing the bg scroll. Also adding baren falls

* Removing baren falls background vibration

* Fixing typo in baren falls background palette change for frm

* Changing events flashes from no-ops to flash-none to avoid any CPU cycle changes

* Moving Remove Flashes flags from misc to graphics, so that they don't affect the seed

* Adding -wmhc option to improve visibility of minimap

* Making location indicator for wmhc stand out even more

* Adding more event flashes to frw

* Changing wmhc based on color feedback

* Replacing boss death flash removal with flash of monster sprite

* switching to Osteoclave's pixel remaster inspired minimap

asilverthorn committed:
Feature: Adding flags to improve Sketch/Control (#19)

* Adding flag to make Sketch/Control 100% effective against valid targets

* fixing minor typo

* Refactoring data/sketch_control.py into separate files. Adding better abilities for Sketch & Control. Making Sketch use caster's stats.

* Minor change to option logging

* Modification to sketch custom command; split into multiple flags; removing control custom commands

* Expanding sia to also include control. Separating Controls from Coliseum monster table.

* Minor arg desc change

* Control uses controller's stats, fix Control bug with Dance moves

* Removing ultima from sketch magimaster

asilverthorn committed:
Feature: Kielbasiago's Adding flag for giving starting Espers (#18)

* add --starting-espers-random flag
- this will give you between the selected range of espers from the very start

* fix err

* Add better esper log

* chore: update flag from sespr => stesp

* chore: remove commented code

* Remove "Random" from "Starting Espers" menu

* Update starting_espers MAX to account for logic issues

* ws, help updated

Co-authored-by: Kiel <95580337+kielbasiago@users.noreply.github.com>

asilverthorn committed:
Feature: Kielbasiago's Update ability to shuffle/mix both dragons and statues (#17)

* add -bmbs flag: adds statues to shuffle pool

* add context for statue ids being in EnemyPacks

* set default true ffor backwards compatibility

* remove default as no way to turn it off

* wip refactor shuffle to "boss location" flags

* bug fixes

* add boilerplate for adding future checks

* chore: no longer appnd dragon locations to flags automatically

* chore: add EOL

* chore: remove unused ids from EnemyPacks

* remove dbugpy code

* Clean up       enemy_packs.randomize_event_bosses

* fix when original bosses with mixed dragon/statues

* now exclude statues when not mix from normal encs

* add KT bosses to condition checks (#2)

Preview: https://youtu.be/wgBJHPq3p6o (Just me clearing KT while playing at 200-400%)
 
Added the following KT boss locations as objective conditions:
- Kefka's Tower Ambush
  - Inferno
  - Bit 59
- Kefka's Tower Guardian
  - Guardian
  - Bit 60
- KT Left Triad Statue
  - Doom
  - Bit 61
- KT Mid Triad Statue
  - Poltergeist
  - Bit 62
- KT Right Triad Statue
  - Goddess
  - Bit 63


## Testing
Used the following flags to test the five encounters: `-oa 40.1.1.11.59 -ob 40.1.1.11.60 -oc 40.1.1.11.61 -od 40.1.1.11.62 -oe 40.1.1.11.63`

* fix initialization of exclude_bosses

Co-authored-by: Nolan <10077353+nolanlocke@users.noreply.github.com>
Co-authored-by: Kiel <95580337+kielbasiago@users.noreply.github.com>

asilverthorn committed:
Feature: Magic MP Randomization (#15)

* Adding Ultima 255 MP flag, Adding Franklin's Magic MP Randomization, Fixing display bugs with 3 digit MP, letting Lore MP randomization go to 255 MP

* Allowing random esper MP cost to go up to 255

* Switching max to 254 to avoid boundary conditions

* making 254 max to avoid issues with MP checks

* Moving from c3 to f0

asilverthorn committed:
Adding --starting-level flag (#14)

asilverthorn committed:
QoL: Adding Rage Move descriptions to rage menu (#13)

* Initial commit of Rage ability names in menu

* Adding logic to display Special rage names

* Adding logic for Special (0xef) attacks

* Rewrite of Rage Description to use custom string table that includes more move details

* More details in rage desc; handling special cases

* Simplifying rage strings

asilverthorn committed:
Feature: Add Remove learnable spells flag + Submenus (#12)

* All original remove learnable spells changes

* Making Life 3 learnable in permadeath seeds

* Fixing bug where Natural Magic users can learn Fire even if excluded

* correcting natural magic learning with excluded commands

* Fixing seed gen error when there's not enough spells available for randomization

* Fixing possibility for long-branches with many submenus

asilverthorn committed:
Removing excluded non-S tier items from tiered or scaled chests (#11)

asilverthorn committed:
Ensuring that Gau can use Magic in FT (#10)

asilverthorn committed:
Feature: Adding random-encounters-chupon flag (#8)

* Adding random-encounters-chupon flag

* Making Chupon 64 sneeze always target entire party

asilverthorn committed:
Feature: Kielbasiago's add --start-junk flag (#7)

* add --start-junk flag

* Add relics to starting junk

Co-authored-by: Kiel <95580337+kielbasiago@users.noreply.github.com>

asilverthorn committed:
QoL: Setting default config options to most commonly used values (#6)

* adding comments

* setting default back to Reset and adding Config 2 relocation for compatibility with DoctorDT tool

asilverthorn committed:
Making Party warp to Arvis house following K@N (#4)

asilverthorn committed:
Feature: Adding -llr flag to randomize L.x lore levels (#3)

* Adding -llr flag to randomize L.x lore levels

* Updating lore battle dialogs

asilverthorn committed:
QoL: Adding clarification of quantity of objective conditions required (#1)

* Adding clarification of quantity of objective conditions required

* responding to feedback on objective menu mods

ff6wc committed:
Update version.py

v1.1.0 -> v.1.2.0

ff6wc committed:
Update flags.py

Updating to match dev branch

ff6wc committed:
Update spells.py

Updated to match dev branch

asilverthorn committed:
Feature: Making Top 4 Magitek commands an objective result (#21)

* Making Terra's Magitek commands an objective result

* Minor changes to address PR comments

Co-authored-by: Will Jones <wrjones104@users.noreply.github.com>

asilverthorn committed:
Feature: Adding warp-all flag for 0 cost starting Warp (#9)

* Adding warp-all flag for 0 cost starting Warp. Refactor ScanAll class to InitialSpells

* Adding missed flag to log

Co-authored-by: Will Jones <wrjones104@users.noreply.github.com>

asilverthorn committed:
Adding -hf flag to hide flags for fun mystery seeds (#2)

Co-authored-by: Will Jones <wrjones104@users.noreply.github.com>
